compiler/rustc_mir_transform/src/value_numbering.rs

```diff
@@ -0,0 +1,907 @@
+struct VnLocal {
+    // The index of the MIR local.
+    local: Local,
+    // This starts at 0 and is bumped whenever we detect a possible mutation to the local.
+    live_range_index: usize,
+}
+
+// Information that we track about each local.
+#[derive(Default)]
+struct VnLocalInfo {
+    // The index of the last statement that potentially mutated this local.
+    statement_index_of_last_kill: usize,
+    // This starts at 0 and is bumped whenever we detect a possible mutation to the local.
+    current_live_range_index: usize,
+}
+
+impl<'tcx> ValueNumberer<'tcx> {
+    fn new(
+        tcx: TyCtxt<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        block_index: BasicBlock,
+    ) -> ValueNumberer<'tcx> {
+        ValueNumberer {
+            tcx,
+            param_env,
+            values: FxHashMap::default(),
+            locals: FxHashMap::default(),
+            location: Location { block: block_index, statement_index: 0 },
+        }
+    }
+
+    // Gathers up all locals and initializes VnLocalInfo for them.
+    fn init_locals(&mut self, block_index: BasicBlock, block: &mut BasicBlockData<'tcx>) {
+        let mut visitor = LocalInitVisitor { locals: FxHashMap::default() };
+        visitor.visit_basic_block_data(block_index, block);
+        self.locals = visitor.locals;
+
+        struct LocalInitVisitor {
+            locals: FxHashMap<Local, VnLocalInfo>,
+        }
+
+        impl<'tcx> Visitor<'tcx> for LocalInitVisitor {
+            fn visit_local(&mut self, local: &Local, _: PlaceContext, _: Location) {
+                if let Entry::Vacant(entry) = self.locals.entry((*local).clone()) {
+                    entry.insert(VnLocalInfo {
+                        statement_index_of_last_kill: 0,
+                        current_live_range_index: 0,
+                    });
+                }
+            }
+        }
+    }
+
+    fn process_statements(
+        &mut self,
+        block: &mut BasicBlockData<'tcx>,
+        local_decls: &LocalDecls<'tcx>,
+    ) {
+        for (statement_index, statement) in block.statements.iter_mut().enumerate() {
+            self.location.statement_index = statement_index;
+
+            match statement.kind {
+                StatementKind::Assign(box (ref place, ref mut rvalue)) => {
+                    self.process_assignment(place, rvalue, &statement.source_info, local_decls);
+                }
+                _ => {
+                    // Non-assignment statements are only analyzed for live range kills.
+                    VnAliasAnalysis::new(local_decls, &mut self.locals)
+                        .visit_statement(statement, self.location)
+                }
+            }
+        }
+    }
+
+    fn process_terminator(&mut self, block: &mut BasicBlockData<'tcx>) {
+        self.location.statement_index = block.statements.len();
+
+        let terminator = match block.terminator {
+            None => return,
+            Some(ref mut terminator) => terminator,
+        };
```

Comment:
Once the MIR is built, `Terminator` is guaranteed to be `Some` (unless you've moved from it elsewhere in this pass). You're free to call `terminator` and `terminator_mut`.